src/transformStateWithClones.js

```diff
@@ -7,7 +7,61 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const result = [];
+
+  for (const action of actions) {
+    const lastItemInArray = result[result.length - 1];
+
+    switch (action.type) {
+      case 'addProperties':
+        let addedProperties;
+
+        if (result.length === 0) {
+          addedProperties = Object.assign({}, state, action.extraData);
+        } else {
+          addedProperties
+            = Object.assign({}, lastItemInArray, action.extraData);
+        }
+
+        result.push(addedProperties);
+        break;
+
+      case 'removeProperties':
+        let removedProperties;
+
+        if (result.length === 0) {
+          removedProperties = Object.assign({}, state);
+        } else {
+          removedProperties = Object.assign({}, lastItemInArray);
+        }
```

Comment:
Same refactoring here + rename variable
```suggestion
        let object;
```